Drive/test/wheel_test.cpp

```diff
@@ -1,75 +1,110 @@
+    CHECK(wheel.GetSteerAngle() == 0);
+  }
+
+  SECTION("Initializing wheel")
   {
     wheel.Initialize();
-    CHECK(wheel.GetSpeed() == doctest::Approx(0.0));
-    CHECK(wheel.GetPosition() == doctest::Approx(0.0));
-    CHECK(wheel.kMaxPosSpeed == 100_rpm);
-    CHECK(wheel.kMaxNegSpeed == -100_rpm);
+    Verify(Method(mock_can, ModuleInitialize)).Twice();
+    CHECK(mock_can.get().CurrentSettings().baud_rate == 1_MHz);
+  }
+
+  SECTION("Setting hub speed to random normal speed")
+  {
+    int random_num = rand() % 100 + 1;
+    wheel.SetHubSpeed(random_num);
+    CHECK(wheel.GetHubSpeed() == random_num);
+
+    random_num = (rand() % 100 - 1) * -1;
+    wheel.SetHubSpeed(random_num);
+    CHECK(wheel.GetHubSpeed() == random_num);
+
+    wheel.SetHubSpeed(0);
+    CHECK(wheel.GetHubSpeed() == 0);
   }
 
-  SECTION("should set wheel speed and angle")
+  SECTION("Setting hub speed beyond maximum speed")
   {
-    wheel.SetHubSpeed(50_rpm);
-    wheel.SetSteeringAngle(20_deg);
-    CHECK(wheel.GetSpeed() == doctest::Approx(50.0));
-    CHECK(wheel.GetPosition() == doctest::Approx(20.0));
-
-    wheel.SetHubSpeed(10_rpm);
-    wheel.SetSteeringAngle(-20_deg);
-    CHECK(wheel.GetSpeed() == doctest::Approx(10.0));
-    CHECK(wheel.GetPosition() == doctest::Approx(0.0));
+    wheel.SetHubSpeed(150);
+    CHECK(wheel.GetHubSpeed() == 100);
+
+    wheel.SetHubSpeed(-150);
+    CHECK(wheel.GetHubSpeed() == -100);
   }
 
-  SECTION("should set extreme wheel speeds")
+  SECTION("Setting steer angle to random normal angle")
   {
-    wheel.SetHubSpeed(150_rpm);
-    CHECK(wheel.GetSpeed() == doctest::Approx(100.0));
+    int random_num = rand() % 360 + 1;
+    wheel.SetSteerAngle(random_num);
+    CHECK(wheel.GetSteerAngle() == random_num);
+
+    random_num = (rand() % 360 - 1) * -1;
+    wheel.SetSteerAngle(random_num);
+    CHECK(wheel.GetSteerAngle() == random_num);
 
-    wheel.SetHubSpeed(-200_rpm);
-    CHECK(wheel.GetSpeed() == doctest::Approx(-100.0));
+    wheel.SetSteerAngle(0);
+    CHECK(wheel.GetSteerAngle() == 0);
   }
 
-  SECTION("should set extreme wheel positions")
+  SECTION("Setting steer angle beyond maximum angle")
   {
-    wheel.SetSteeringAngle(500_deg);
-    CHECK(wheel.GetPosition() == doctest::Approx(360.0));
+    wheel.SetSteerAngle(400);
+    CHECK(wheel.GetSteerAngle() == 40);
 
-    wheel.SetSteeringAngle(-360_deg);  // resets wheel to zero
-    wheel.SetSteeringAngle(-500_deg);
-    CHECK(wheel.GetPosition() == doctest::Approx(-360.0));
+    wheel.SetSteerAngle(-400);
+    CHECK(wheel.GetSteerAngle() == -40);
   }
 
-  SECTION("should home wheel positions")
+  SECTION("Spying on Wheel")
   {
-    wheel.HomeWheel();
-    CHECK(wheel.homing_offset_angle_ == 0_deg);
+    // TODO: Get spy logic working correctly - idk why it's not working...
+    Mock<drive::Wheel> spy(wheel);
+    When(Method(spy, GetSteerAngle)).AlwaysReturn(10);
+    drive::Wheel & test = spy.get();
+    CHECK(test.GetSteerAngle() == 10);
   }
```

Comment:
@kammce I've been trying to figure out how spying works in FakeIt but for whatever reason this does not work. It returns 0 as opposed to AlwaysReturn() = 10

Reply:
FakeIt can only work on virtual functions. In Java all functions are virtual functions which is why you can override any function you like. With C++ a normal function is hard coded and you cannot get the application to redirect to another function address so easily at runtime (should be impossible by principle). So if you wan this mockable by FakeIt, you'll need to make the function virtual.

This was actually one of my inspirations for why SJSU-Dev2 is interface heavy and why almost every driver has to be an interface. So that every driver can have its implementation swapped out on the fly in the unit testing library. 

For ease of use now, and not refactoring everything for better unit testing capability, I'd say, just make the methods you want to test virtual and we will optimize this later. :)

Reply:
ohhh okay makes sense now, thanks will do